### PR TITLE
8265263: AArch64: Combine vneg with right shift count

### PR DESCRIPTION
**Pull request by dchuyko**
This is a performance improvement for AArch64. There are several differences from the original change.

https://bugs.openjdk.org/browse/JDK-8267356 (Vector API SVE codegen support) is not in 17u, so `UseSVE == 0` parts in predicates are missing/excluded.

https://bugs.openjdk.org/browse/JDK-8288445 (C2 compilation fails) is a subsequent bugfix already backported in 17u, so some `immI` arguments in rules became `immI_positive`.

https://bugs.openjdk.org/browse/JDK-8277239 (SIGSEGV in vrshift_reg_maskedNode::emit) is also related to Vector API and is not in 17u, so `!n->as_ShiftV()->is_var_shift()` is replaced by `VectorNode::is_vshift_cnt(n->in(2))`. This substitution may raise doubts.

Testing: jtreg test/hotspot/jtreg/compiler, tier1, tier2 on aarch64.

Performance improvements in the added benchmark VectorShiftRight on Graviton 2 for default size=1024 correspond to the original review:

```
rShiftByte      16%
rShiftInt       27%
rShiftLong      16%
rShiftShort     20%
urShiftByte     0%
urShiftChar     20%
urShiftInt      27%
urShiftLong     16%
```

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8265263](https://bugs.openjdk.org/browse/JDK-8265263): AArch64: Combine vneg with right shift count

### Reviewers
 * [Hao Sun](https://openjdk.org/census#haosun) (@shqking - no project role) ⚠️ Review applies to [0a9c5957](https://git.openjdk.org/jdk17u-dev/pull/811/files/0a9c5957148fba5d81ecf77d5bb67c2241902f88)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-dev pull/811/head:pull/811` \
`$ git checkout pull/811`

Update a local copy of the PR: \
`$ git checkout pull/811` \
`$ git pull https://git.openjdk.org/jdk17u-dev pull/811/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 811`

View PR using the GUI difftool: \
`$ git pr show -t 811`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-dev/pull/811.diff">https://git.openjdk.org/jdk17u-dev/pull/811.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/aarch64_neon.ad [hunk 1]
```diff
@@ -4243,32 +4243,64 @@ instruct vxor16B(vecX dst, vecX src1, vecX src2)
...
+// Vector shift count
+// Note-1: Low 8 bits of each element are used, so it doesn't matter if we
+//         treat it as ints or bytes here.
+// Note-2: Shift value is negated for RShiftCntV additionally. See the comments
+//         on vsra8B rule for more details.
+
+instruct vslcnt8B(vecD dst, iRegIorL2I cnt) %{
   predicate(n->as_Vector()->length_in_bytes() == 4 ||
-            n->as_Vector()->length_in_bytes() == 8);
+                            n->as_Vector()->length_in_bytes() == 8);
```
Comment (shqking) [inline src/hotspot/cpu/aarch64/aarch64_neon.ad:4254]: We may improve the style here, i.e. removing the leading whitespace.

File: src/hotspot/cpu/aarch64/aarch64_neon.ad [hunk 2]
```diff
@@ -4243,32 +4243,64 @@ instruct vxor16B(vecX dst, vecX src1, vecX src2)
...
+  ins_cost(INSN_COST);
   format %{ "dup  $dst, $cnt\t# shift count vector (16B)" %}
   ins_encode %{
     __ dup(as_FloatRegister($dst$$reg), __ T16B, as_Register($cnt$$reg));
   %}
   ins_pipe(vdup_reg_reg128);
 %}
 
+instruct vsrcnt8B(vecD dst, iRegIorL2I cnt) %{
+  predicate(UseSVE == 0 && (n->as_Vector()->length_in_bytes() == 4 ||
```
Comment (shqking) [inline src/hotspot/cpu/aarch64/aarch64_neon.ad:4276]: As mentioned in the commit msg, I think `UseSVE == 0` check is not needed here.

File: src/hotspot/cpu/aarch64/aarch64_neon.ad (hunk 1 above)
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:4254]: Not in a backport.

File: src/hotspot/cpu/aarch64/aarch64_neon.ad (hunk 2 above)
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:4276]: Again, the place to fix that is JDK head.

File: src/hotspot/cpu/aarch64/aarch64_neon.ad (hunk 1 above)
Comment (shqking) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:4254]: I see. Thanks for your comment.

File: src/hotspot/cpu/aarch64/aarch64_neon.ad (hunk 2 above)
Comment (shqking) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:4276]: But `UseSVE == 0` check is removed for `vslcnt8B` and `vslcnt16B` already.
As stated in the commit msg, VectorAPI  SVE codegen part is missing for 17u.

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:4276]: > But `UseSVE == 0` check is removed for `vslcnt8B` and `vslcnt16B` already. As stated in the commit msg, VectorAPI SVE codegen part is missing for 17u.

OK.

Comment (dchuyko) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad]: Thanks, I removed 2 checks symmetrically.

**Review by shqking — APPROVED** [review]
Thanks for your update.
Overall, it looks good to me (I'm not a Reviewer).

**Comment by shqking** [comment]
I noticed that there exists difference between the generated AD file from M4 file and the provided AD file.
I think we should eliminate the difference.

```bash
~/jdk//src/hotspot/cpu/aarch64$ m4 aarch64_neon_ad.m4 > aarch64_neon.ad
~/jdk/src/hotspot/cpu/aarch64$ git diff
 src/hotspot/cpu/aarch64/aarch64_neon.ad | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/hotspot/cpu/aarch64/aarch64_neon.ad b/src/hotspot/cpu/aarch64/aarch64_neon.ad
index db50e08fffd..d43c8d31b78 100644
--- a/src/hotspot/cpu/aarch64/aarch64_neon.ad
+++ b/src/hotspot/cpu/aarch64/aarch64_neon.ad
@@ -4250,8 +4250,8 @@ instruct vxor16B(vecX dst, vecX src1, vecX src2)
 //         on vsra8B rule for more details.

 instruct vslcnt8B(vecD dst, iRegIorL2I cnt) %{
-  predicate(n->as_Vector()->length_in_bytes() == 4 ||
-                            n->as_Vector()->length_in_bytes() == 8);
+  predicate((n->as_Vector()->length_in_bytes() == 4 ||
+              n->as_Vector()->length_in_bytes() == 8));
   match(Set dst (LShiftCntV cnt));
   ins_cost(INSN_COST);
   format %{ "dup  $dst, $cnt\t# shift count vector (8B)" %}
@@ -4273,8 +4273,8 @@ instruct vslcnt16B(vecX dst, iRegIorL2I cnt) %{
 %}

 instruct vsrcnt8B(vecD dst, iRegIorL2I cnt) %{
-  predicate(n->as_Vector()->length_in_bytes() == 4 ||
-            n->as_Vector()->length_in_bytes() == 8);
+  predicate((n->as_Vector()->length_in_bytes() == 4 ||
+             n->as_Vector()->length_in_bytes() == 8));
   match(Set dst (RShiftCntV cnt));
   ins_cost(INSN_COST * 2);
   format %{ "negw  rscratch1, $cnt\t"

```

**Comment by theRealAph** [comment]
> I noticed that there exists difference between the generated AD file from M4 file and the provided AD file. I think we should eliminate the difference.

Yes, I agree. Always.

**Comment by dchuyko** [comment]
> 

Yes, missed that. Updated the branch, trying to keep leading whitespace in .m4 close to the original diff.

**Review by shqking — APPROVED** [review]
LGTM. (I'm not a Reviewer)

**Review by phohensee — CHANGES_REQUESTED** [review]
I'm not an expert on this code, but I don't think VectorNode::is_vshift_cnt(n->in(2)) is equivalent to !n->as_ShiftV()->is_var_shift(). is_vshift_cnt() just checks for LShiftCntV and RShiftCntV
```
bool VectorNode::is_vshift_cnt(Node* n) {
  switch (n->Opcode()) {
  case Op_LShiftCntV:
  case Op_RShiftCntV:
    return true;
  default:
    return false;
  }
}
```
but these operators can have variable shift counts, see in aarch64_sve.ad
```
instruct vshiftcntB(vReg dst, iRegIorL2I cnt) %{
  predicate(UseSVE > 0 && n->as_Vector()->length() >= 16 &&
            (n->bottom_type()->is_vect()->element_basic_type() == T_BYTE));
  match(Set dst (LShiftCntV cnt));
  match(Set dst (RShiftCntV cnt));
  format %{ "sve_dup $dst, $cnt\t# vector shift count (sve) (B)" %}
  ins_encode %{
    __ sve_dup(as_FloatRegister($dst$$reg), __ B, as_Register($cnt$$reg));
  %}
  ins_pipe(pipe_slow);
%}
```

**Comment by dchuyko** [comment]
I consider backporting of JDK-8277239, however the test mentioned there is stable on 17u in my experiments and the change doesn't apply cleanly. And there is also arm32 follow-up (JDK-8278267) that may be considered.

**Pull request closed without merging**